src/Common/SourceControl/UnityVersionControl/UnityVersionControlFileHashProvider.cs

```diff
@@ -0,0 +1,196 @@
+        {
+            _logger = logger;
+        }
+        public async Task<IReadOnlyDictionary<string, byte[]>> GetFileHashesAsync(string repoRoot, CancellationToken cancellationToken)
+        {
+            Task<Dictionary<string, byte[]>> hashesTask = GetRepoFileHashesAsync(repoRoot, cancellationToken);
+            return await hashesTask;
+        }
+
+        private async Task<Dictionary<string, byte[]>> GetRepoFileHashesAsync(string basePath, CancellationToken cancellationToken)
+        {
+            return await UnityVersionControl.RunAsync(_logger, workingDir: basePath, "ls -R --format=\"{path}\t{hash}\"",
+                (_, stdout) => Task.Run(() => ParseUnityLsFiles(stdout, (filesToRehash, fileHashes) => GitHashObjectAsync(basePath, filesToRehash, fileHashes, cancellationToken))),
+                (exitCode, result) =>
+                {
+                    if (exitCode != 0)
+                    {
+                        throw new SourceControlHashException("cm ls failed with exit code " + exitCode);
+                    }
+
+                    return result;
+                },
+                cancellationToken);
+        }
+
+        internal async Task<Dictionary<string, byte[]>> ParseUnityLsFiles(
+        TextReader cmOutput,
+        Func<List<string>, Dictionary<string, byte[]>, Task> hasher)
+        {
+            // relativePathInRepository<tab>hash
+            using var reader = new UnityVersionContorlLsFileOutputReader(cmOutput);
+            var fileHashes = new Dictionary<string, byte[]>(StringComparer.OrdinalIgnoreCase);
+            var filesToRehash = new List<string>();
+            StringBuilder? line;
+            while ((line = reader.ReadLine()) != null)
+            {
+                var splitLine = line.ToString().Split('\t');
+                string file = splitLine[0];
+                if (splitLine.Length > 1 && splitLine[1].Length > 0)
+                {
+                    string hash = splitLine[1];
+                    fileHashes[file] = HexUtilities.Base64ToBytes(hash);
+                }
+                else
+                {
+                    _logger.LogMessage($"{file} is missing a hash and will be rehashed.");
+                    filesToRehash.Add(file);
+                }
+            }
+
+            if (filesToRehash.Count > 0)
+            {
+                Stopwatch sw = Stopwatch.StartNew();
+                // we could do this as new files come in just not clear it's worth it
+                await hasher(filesToRehash, fileHashes);
+                _logger.LogMessage($"{fileHashes.Count} files Rehashing {filesToRehash.Count} modified files took {sw.ElapsedMilliseconds} msec");
+            }
+
+            return fileHashes;
+        }
+
+        internal Task GitHashObjectAsync(string basePath, List<string> filesToRehash, Dictionary<string, byte[]> filehashes, CancellationToken cancellationToken)
+        {
+            return Git.RunAsync(
```

Comment:
It seems weird to fall back to git here. Does Unity not have an equivalent command to use?